crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -2071,7 +2062,10 @@ impl<'a> Checker<'a> {
 
                     self.semantic.restore(snapshot);
 
-                    if self.semantic.in_annotation() && self.semantic.future_annotations_or_stub() {
+                    if self.semantic.in_annotation()
+                        && (self.semantic.in_typing_only_context()
+                            || self.semantic.future_annotations())
+                    {
```

Comment:
I think honestly PYI020 is a better fit for stub files and `if TYPE_CHECKING` blocks, so the logic here should probably be changed. But for now I'm just maintaining compatibility with previous versions of Ruff, which emitted this lint on annotations in stub files as well as annotations in `.py` files. Abruptly reducing the scope of the rule like that would maybe be slightly backwards-incompatible; it at least should be considered in a separate PR.